components/script/dom/gpubuffer.rs

```diff
@@ -313,24 +314,21 @@ impl GPUBufferMethods for GPUBuffer {
             return Err(Error::Operation);
         }
 
-        unsafe extern "C" fn free_func(_contents: *mut c_void, free_user_data: *mut c_void) {
-            let _ = Rc::from_raw(free_user_data as _);
-        }
+        let vec = m_info.mapping.borrow().to_vec();
 
-        let array_buffer = unsafe {
-            NewExternalArrayBuffer(
-                *cx,
-                range_size as usize,
-                m_info.mapping.borrow_mut()[offset as usize..m_end as usize].as_mut_ptr() as _,
-                Some(free_func),
-                Rc::into_raw(m_info.mapping.clone()) as _,
```

Comment:
So this is actually quite simple: 

1. This actually works, because it clones the Rc(increments the count), before calling `into_raw`. 
2. I think we should switch to an `Arc<Mutex<`, because as [per the docs](https://searchfox.org/mozilla-central/rev/2a867dd1ab015c3ef24b774a57709fb3b3dc4961/js/public/ArrayBuffer.h#169) calling `free_func` should be threadsafe, which the rc is not. 

That's it. So we just have to move the function call to the bindings, and set the internal of the typed array to be the js object returned by `NewExternalArrayBuffer`. 

[here](https://github.com/servo/servo/pull/31202/files#diff-caa3a7eda41967ceeede122072e749d9bda074ea1860d63c862817b61bc6baf0R299) we should return an `Err(Error::Operation)` if `m_info` is none.